components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/api/auth/ApiAuthnEndpoint.java

```diff
@@ -0,0 +1,219 @@
+import org.wso2.carbon.identity.oauth.endpoint.authz.OAuth2AuthzEndpoint;
+import org.wso2.carbon.identity.oauth.endpoint.exception.InvalidRequestParentException;
+
+import java.net.URISyntaxException;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.Response;
+
+/**
+ * Class containing the REST API for API based authentication.
+ */
+@Path("/authn")
+public class ApiAuthnEndpoint {
+
+    private final AuthenticationService authenticationService = new AuthenticationService();
+    private final OAuth2AuthzEndpoint oAuth2AuthzEndpoint = new OAuth2AuthzEndpoint();
+    private static final String AUTHENTICATOR_IDP_SPLITTER = ":";
+    private static final String AUTHENTICATOR = "authenticator";
+    private static final String IDP = "idp";
+    private static final Log LOG = LogFactory.getLog(ApiAuthnEndpoint.class);
+    private static final ApiAuthnHandler API_AUTHN_HANDLER = new ApiAuthnHandler();
+
+    @POST
+    @Path("/")
+    @Consumes("application/json")
+    @Produces("application/json")
+    public Response handleAuthentication(@Context HttpServletRequest request, @Context HttpServletResponse response,
+                                         String payload) {
+
+        try {
+            AuthRequest authRequest = buildAuthRequest(payload);
+            AuthServiceRequest authServiceRequest = getAuthServiceRequest(request, response, authRequest);
+            AuthServiceResponse authServiceResponse = authenticationService.handleAuthentication(authServiceRequest);
+
+            switch (authServiceResponse.getFlowStatus()) {
+                case INCOMPLETE:
+                    return handleIncompleteAuthResponse(authServiceResponse);
+                case SUCCESS_COMPLETED:
+                    return handleSuccessCompletedAuthResponse(request, response, authServiceResponse);
+                case FAIL_INCOMPLETE:
+                    return handleFailAuthResponse(authServiceResponse);
+                case FAIL_COMPLETED:
+                    return handleFailAuthResponse(authServiceResponse);
+                default:
+                    throw new AuthServiceException("Unknown flow status: " + authServiceResponse.getFlowStatus());
+            }
+
+        } catch (AuthServiceClientException e) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Client error while handling authentication request", e);
+            }
+            return buildOAuthInvalidRequestError(e.getMessage());
+        } catch (AuthServiceException | InvalidRequestParentException | URISyntaxException e) {
```

Comment:
Is  InvalidRequestParentException  a server error? The name of the class sounds different.

Reply:
its a client error. fixed.